wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,370 @@
+   * URL, a monitor for that custom endpoint will be created if it does not already exist.
+   *
+   * @param hostSpec    The host information for the desired connection.
+   * @param props       The connection properties.
+   * @param connectFunc The connect pipeline function to call to establish the connection.
+   * @return The connection established by the {@code connectFunc}.
+   * @throws SQLException If an error occurs while attempting to establish a connection, or if we are connecting to a
+   *                      custom endpoint and we time out while waiting for custom endpoint info to be placed in the
+   *                      cache by the custom endpoint monitor.
+   */
+  protected Connection connectInternal(
+      HostSpec hostSpec,
+      Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    if (!this.rdsUtils.isRdsCustomClusterDns(hostSpec.getHost())) {
+      return connectFunc.call();
+    }
+
+    this.customEndpointHostSpec = hostSpec;
+    this.endpointConnectionsCounter.inc();
+    LOGGER.finest(
+        Messages.get(
+            "CustomEndpointPlugin.connectionRequestToCustomEndpoint", new Object[]{ hostSpec.getHost() }));
+
+    createMonitorIfAbsent(props);
+
+    // If needed, wait a short time for custom endpoint info to be discovered.
+    boolean isInfoInCache = waitForCustomEndpointInfo();
+
+    if (!isInfoInCache) {
+      throw new SQLException(
+          Messages.get("CustomEndpointPlugin.cacheTimeout",
+              new Object[]{this.waitOnCachedInfoDurationMs, this.customEndpointHostSpec.getHost()}));
+    }
+
+    return connectFunc.call();
+  }
+
+  /**
+   * Creates a monitor for the custom endpoint if it does not already exist.
+   *
+   * @param props The connection properties.
+   */
+  protected void createMonitorIfAbsent(Properties props) {
+    monitors.computeIfAbsent(
+        this.customEndpointHostSpec.getHost(),
+        (customEndpoint) -> new CustomEndpointMonitorImpl(
+            this.pluginService,
+            this.customEndpointHostSpec,
+            getRegion(this.customEndpointHostSpec, props),
+            TimeUnit.MILLISECONDS.toNanos(CUSTOM_ENDPOINT_INFO_REFRESH_RATE_MS.getLong(props)),
+            this.rdsClientFunc
+        ),
+        TimeUnit.MILLISECONDS.toNanos(this.idleMonitorExpirationMs)
+    );
+  }
+
+  /**
+   * Determines the AWS region from the given parameters. The region specified by the props will be used if it is
+   * provided. Otherwise, the region will be parsed from the specified host URL.
+   *
+   * @param hostSpec Host information for the connection being established.
+   * @param props    The connection properties for the connection being established.
+   * @return The AWS region of the connection being established.
+   */
+  protected Region getRegion(HostSpec hostSpec, Properties props) {
```

Comment:
I've seen similar piece of code before. It seems we can put it in a helper class.